### PR TITLE
minimal viable fix for insecure constantize

### PR DESCRIPTION
**Pull request by zvkemp**
Let's prevent miscreants from crashing the server!

- [x] only allow validatable models to be constantized
- [x] provide a whitelist of classes available for constantization

@lumoslabs/platform

### CODE REVIEW

**Comment by zvkemp** [comment]
@jdoconnor

**Comment by jdoconnor** [comment]
:balloon:

File: lib/comply/whitelist_constantize.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+module Comply
+  module WhitelistConstantize
+    class NotWhitelistedError < ArgumentError
+    end
+
+    DEFAULT_PROC = -> (klass) { klass <= ActiveModel::Validations }
+
+    class << self
+      def constantize(str)
+        whitelist! { str.classify.constantize }
```
Comment (camertron) [inline lib/comply/whitelist_constantize.rb]: Why not just call this method with an argument? The block/yield isn't necessary.

File: lib/comply/whitelist_constantize.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+      def whitelist
+        @whitelist ||= [DEFAULT_PROC]
+      end
+
+      private
+
+      def whitelist!
+        yield.tap do |klass|
+          whitelist.detect(-> { raise NotWhitelistedError }) do |matcher|
```
Comment (camertron) [inline lib/comply/whitelist_constantize.rb]: I had no idea until now that detect accepts an `ifnone` argument - could you add a comment?

File: lib/comply/whitelist_constantize.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    end
+
+    DEFAULT_PROC = -> (klass) { klass <= ActiveModel::Validations }
+
+    class << self
+      def constantize(str)
+        whitelist! { str.classify.constantize }
+      end
+
+      def whitelist
```
Comment (camertron) [inline lib/comply/whitelist_constantize.rb]: Do you add an additional whitelist element by doing `WhitelistConstantize.whitelist << -> { ... }`? Might be nice to have a method like `add_to_whitelist` or something to encapsulate this behavior.

File: lib/comply/whitelist_constantize.rb (hunk 1 above)
Comment (zvkemp) [inline lib/comply/whitelist_constantize.rb]: Good call

**Comment by zvkemp** [comment]
@camertron how does this look?

**Comment by azach** [comment]
seems like this should be a major version bump since it's backwards incompatible with the previous versions?

**Comment by zvkemp** [comment]
@azach It's *technically* not breaking anything that wasn't already broken. The implementation of the controller used the `valid?` and `error` methods provided from ActiveModel; this update makes that expectation explicit rather than implicit.

**Comment by zvkemp** [comment]
I'll test it with lumos_rails and if any changes are required I'll bump the version up.

**Comment by camertron** [comment]
All legitimate implementations of this in lumos_rails should be on classes that inherit from `ActiveModel::Validations`, which this PR checks for by default (it's the only check in the whitelist), which means it won't (read: shouldn't) break anything.

Looks good to me, @zvkemp :+1:

**Comment by azach** [comment]
gotcha, makes sense to me :+1:

**Pull request merged**